src/main/java/org/folio/rest/util/RestUtil.java

```diff
@@ -0,0 +1,179 @@
+import io.vertx.ext.web.client.WebClient;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.folio.HttpStatus;
+import org.folio.rest.tools.utils.VertxUtils;
+
+import javax.ws.rs.BadRequestException;
+import javax.ws.rs.InternalServerErrorException;
+import javax.ws.rs.NotFoundException;
+import java.util.Map;
+
+import static org.folio.HttpStatus.HTTP_OK;
+import static org.folio.HttpStatus.HTTP_CREATED;
+import static org.folio.HttpStatus.HTTP_NOT_FOUND;
+import static org.folio.HttpStatus.HTTP_NO_CONTENT;
+import static org.folio.HttpStatus.HTTP_INTERNAL_SERVER_ERROR;
+
+public final class RestUtil {
+
+  private static final Logger LOGGER = LogManager.getLogger();
+  private static final String STATUS_CODE_IS_NOT_SUCCESS_MSG = "Response HTTP code is not equals 200, 201, 204. Response code: {}";
+
+  public static class WrappedResponse {
+    private int code;
+    private String body;
+    private JsonObject json;
+    private HttpResponse<Buffer> response;
+
+    WrappedResponse(int code, String body,
+                    HttpResponse<Buffer> response) {
+      this.code = code;
+      this.body = body;
+      this.response = response;
+      try {
+        json = new JsonObject(body);
+      } catch (Exception e) {
+        json = null;
+      }
+    }
+
+    public int getCode() {
+      return code;
+    }
+
+    public String getBody() {
+      return body;
+    }
+
+    public HttpResponse<Buffer> getResponse() {
+      return response;
+    }
+
+    public JsonObject getJson() {
+      return json;
+    }
+  }
+
+  private RestUtil() {
+  }
+
+  /**
+   * Create http request
+   *
+   * @param url     - url for http request
+   * @param method  - http method
+   * @param payload - body of request
+   * @return - async http response
+   */
+  public static <T> Future<WrappedResponse> doRequest(OkapiConnectionParams params, String url,
+                                                      HttpMethod method, T payload) {
+    Promise<WrappedResponse> promise = Promise.promise();
+    try {
+      Map<String, String> headers = params.getHeaders();
+      String requestUrl = params.getOkapiUrl() + url;
+      WebClient client = WebClient.wrap(getHttpClient(params));
+
+      HttpRequest<Buffer> request = client.requestAbs(method, requestUrl);
+      if (headers != null) {
+        headers.put("Content-type", "application/json");
+        headers.put("Accept", "application/json, text/plain");
+        for (Map.Entry<String, String> entry : headers.entrySet()) {
+          request.putHeader(entry.getKey(), entry.getValue());
+        }
+      }
+      if (method == HttpMethod.PUT || method == HttpMethod.POST) {
+        request.sendBuffer(Buffer.buffer(new ObjectMapper().writeValueAsString(payload)), handleResponse(promise));
+      } else {
+        request.send(handleResponse(promise));
+      }
+      return promise.future();
+    } catch (Exception e) {
+      promise.fail(e);
```

Comment:
Exception Fail reason not captched in method RestUtil.doRequest(params, "/pubsub/publish", HttpMethod.POST, event).
As an option we can add log here

Reply:
Done